server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/model/MockSMTPBehaviorInformation.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    static class UnlimitedAnswersCounter implements RemainingAnswersCounter {
+
+        @Override
+        public void decrease() {
+            // do nothing
+        }
+
+        @Override
+        public boolean hasRemainingAnswers() {
+            return true;
+        }
+
+        @Override
+        public Optional<Integer> getValue() {
+            return Optional.empty();
+        }
+    }
+
+    static class AnswersCounter implements RemainingAnswersCounter {
+
+        static AnswersCounter remains(int remainingCount) {
+            return new AnswersCounter(remainingCount);
+        }
+
+        private final AtomicInteger remainingCount;
+
+        AnswersCounter(int remainingCount) {
+            Preconditions.checkArgument(remainingCount > 0, "remainingCount should be positive");
+
+            this.remainingCount = new AtomicInteger(remainingCount);
+        }
+
+        @Override
+        public void decrease() {
+            remainingCount.updateAndGet(currentCount -> {
+                Preconditions.checkState(currentCount > 0,
+                    "remainingCount is already being zero, you can not decrease more");
+
+                return currentCount - 1;
+            });
+        }
+
+        @Override
+        public boolean hasRemainingAnswers() {
+            return remainingCount.get() > 0;
+        }
+
+        @Override
+        public Optional<Integer> getValue() {
+            return Optional.of(remainingCount.get());
+        }
+    }
+
+    public static MockSMTPBehaviorInformation from(MockSMTPBehavior behavior) {
+        return new MockSMTPBehaviorInformation(
+            behavior,
+            RemainingAnswersCounter.from(behavior.getNumberOfAnswer()));
+    }
+
+    private final MockSMTPBehavior behavior;
+    private final RemainingAnswersCounter remainingAnswersCounter;
+
+    MockSMTPBehaviorInformation(MockSMTPBehavior behavior, RemainingAnswersCounter remainingAnswersCounter) {
+        Preconditions.checkNotNull(behavior);
+        Preconditions.checkNotNull(remainingAnswersCounter);
+
+        this.behavior = behavior;
+        this.remainingAnswersCounter = remainingAnswersCounter;
+    }
+
+    public void decreaseRemainingAnswers() {
+        remainingAnswersCounter.decrease();
+    }
+
+    public MockSMTPBehavior getBehavior() {
+        return behavior;
+    }
+
+    public boolean hasRemainingAnswers() {
+        return remainingAnswersCounter.hasRemainingAnswers();
+    }
+
+    @VisibleForTesting
+    public Optional<Integer> remainingAnswersCounter() {
+        return remainingAnswersCounter.getValue();
+    }
+
+    @Override
+    public final boolean equals(Object o) {
+        if (o instanceof MockSMTPBehaviorInformation) {
+            MockSMTPBehaviorInformation that = (MockSMTPBehaviorInformation) o;
+
+            return Objects.equals(this.behavior, that.behavior)
+                && Objects.equals(this.remainingAnswersCounter(), that.remainingAnswersCounter());
```

Comment:
Again, this don't make sense to do equals on counter (2 counters are always different). I think this equal + hashcode have no meaning and should be removed.

Reply:
Acutally I compare two counters in Integer but not two counter objects. `this.remainingAnswersCounter()` returns an `Optional<Integer>`